src/v/config/constraint_store.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/configuration.h"
+#include "config/constraint_types.h"
+
+#include <fmt/format.h>
+
+#include <unordered_map>
+
+namespace config {
+// Similar to config_store
+class constraint_store {
+public:
+    ~constraint_store() = default;
+
+    bool contains(ss::sstring& name) { return _constraints.contains(name); }
+
+    constraint& get(const ss::sstring& name) {
+        auto found = _constraints.find(name);
+        if (found != _constraints.end()) {
+            return found->second;
+        } else {
+            throw std::out_of_range(
+              fmt::format("Constraint {} not found", name));
+        }
+    }
+
+    /**
+     * Creates constraints based on the constraint mapping from the
+     * configuration store. This implies that the broker already read the YAML
+     * config file.
+     */
+    void load_constraints() {
+        load_constraints(
+          config::shard_local_cfg().constraints(), config::shard_local_cfg());
+    }
+
+    void load_constraints(
+      const std::unordered_map<constraint_args::key_type, constraint_args>&
+        constraint_args,
+      config::config_store& conf) {
+        _constraints.reserve(constraint_args.size());
+        for (const auto& [name, args] : constraint_args) {
+            auto& property = conf.get(name);
+            constraint cons{
+              std::move(args), property.make_constraint_methods()};
+            _constraints[name] = std::move(cons);
+        }
+    }
+
+    template<typename Func>
+    void for_each(Func&& f) const {
+        for (const auto& [_, cons] : _constraints) {
+            f(cons);
+        }
+    }
+
+private:
+    std::unordered_map<std::string_view, constraint> _constraints;
```

Comment:
> A btree

The idea was to leverage the search and insert runtime advantages that unordered_map has over map and btree impls. Also would not need to check for duplicates.

> A flat_hash_map

The benefits to flat_hash_map are memory efficiency and invalidation to iterators, references, and pointers such as upon rehash. This impl does not store the key or value as references or pointers so we can ignore those.

The motivation for this project is to assist users with ensuring data safety on Kafka topics so the expectation is that constraints are set infrequently: at startup from reading the YAML config file or with the Admin API. Unless a user is invoking the Admin API to dynamically change constraints in response to events, which is a strange use-case, the benefits for memory efficiency and iterator invalidation are negligible.

With that said, a possible corner case for frequent changes with the Admin API is at early feature adoption because users are figuring out optimal constraints. This also implies that, after some time, constraints are unchanged.

Happy to chat more offline if this requires further discussion.

Reply:
After giving this more thought, a flat_hash_map is better since the map is rehashed when loading constraints.